jp2a.1
.TH jp2a 1  "July 12, 2006" "version 0.9.14" "USER COMMANDS"
.SH NAME
jp2a \- convert JPEG images to ASCII characters
.SH SYNOPSIS
.B jp2a
[
.I options
] [
.I filename(s) | URL(s)
]
.LP
.SH DESCRIPTION
.B jp2a
will convert JPEG images to ASCII characters.  You can specify a mixture of files and URLs.
.SH OPTIONS
.TP
.BI \-
Read JPEG image from standard input
.TP
.B \-b \-\-border
Frame output image in a border
.TP
.BI \-\-chars= ....
Use the given characters when producing the output ASCII image
.TP
.B \-d \-\-debug
Print debugging information when using libcurl to download images from the net.
.TP
.B \-\-flipx
Flip output image horizontally
.TP
.B \-\-flipy
Flip output image vertically
.TP
.BI \-\-height= N
Set output height.  If only --height is specified, then output width will
be calculated according to the source images aspect ratio.
.TP
.B \-h \-\-help
Display a short help text
.TP
.B \-\-html
Make ASCII output in strict XHTML 1.0, suitable for viewing with web browsers.

This is useful with big output dimensions, and you want to check the result
with a browser with small font.

See also \-\-html-fontsize.
.TP
.BI \-\-html\-fontsize= N
Set fontsize when using \-\-html output.  Default is 4.
.TP
.B \-i \-\-invert
Invert output image.  This is very useful for example if you view a picture
where white is the background color, and you're using light characters on a
dark (black) display.
.TP
.BI \-\-size= WIDTHxHEIGHT
Set output dimension.
.TP
.B \-v \-\-verbose
Print some verbose information to standard error when reading each JPEG image.
.TP
.BI \-\-width= N
Set output width.  If you only specify the width, the height will be
calculated automatically.  Default mode is --width=78.
.TP
.B \-V \-\-version
Print program version.
.SH DOWNLOADING IMAGES FROM THE NET
.TP
If you have compiled jp2a with libcurl(3), you can download images by specifying URLs:

.B jp2a
https://user:pass@foo.com/bar.jpg
.PP
The protocols recognized are ftp, ftps, file, http, https and tftp.
.PP
If you need more control of the downloading, you should use curl(1) or wget(1) and jp2a
read the image from standard input.
.PP
jp2a uses pipe and fork to download images using libcurl (i.e., no exec or system calls) 
and therefore does not worry about malevolently formatted URLs.
.SH EXAMPLES
.TP
Convert and print imagefile.jpg using ASCII characters in 40 columns and 20 rows:

.B jp2a
\-\-size=40x20 imagefile.jpg
.TP
Download an image off the net, convert and print:

.B jp2a
http://www.google.com/intl/en/logos/easter_logo.jpg
.TP
Output picture.jpg and picture2.jpg, each 80x25 characters, using the characters " ...ooxx@@" for output:

.B jp2a
\-\-size=80x25 --chars=" ...ooxx@@" picture.jpg picture2.jpg
.TP
Output image.jpg using 76 columns, height is automatically calculated from aspect ratio of image.jpg

cat image.jpg |
.B jp2a
\-\-width=76 \-
.TP
If you use jp2a together with ImageMagick's convert(1) then you can make good use of pipes, and have ImageMagick do all sorts of image conversions and effects on the source image.  For example:

.B convert
somefile.png jpg:\- |
.B jp2a
\- \-\-width=80
.TP
Check out convert(1) options to see what you can do.  Convert can handle almost any image format, so with this combination you can convert images in e.g. PDF or AVI files to ASCII.
.TP
Although the default build of jp2a includes automatic downloading of files specified by URLs, you can explicitly
download them by using curl(1) or wget(1), for example:

.B curl
\-s
http://foo.bar/image.jpg |
.B convert
\- jpg:- |
.B jp2a
\-
.SH EXIT STATUS
jp2a returns zero for normal operations, 1 when errors are encountered.
.SH PROJECT HOMEPAGE
The latest version of jp2a and news is always available from

http://jp2a.sf.net
.SH AUTHOR
Christian Stigen Larsen <csl@sublevel3.org> -- http://csl.sublevel3.org

jp2a uses jpeglib to read JPEG files.  jpeglib is made by The Independent JPEG Group (IJG),
who have a page at http://www.ijg.org

.SH BUGS
The program does not interpolate when resizing the JPEG image, and thus
big output dimensions will not give very accurate looking results.  If you
want better quality, you might be able to use convert(1) to convert the source
image to the exact ASCII output dimensions of jp2a.
.SH SEE ALSO
cjpeg(1), djpeg(1), jpegtran(1), convert(1)


NEWS
Version 0.9.0
        - First public release


jp2a.cpp
/*
 * Copyright (C) 2006 Christian Stigen Larsen <http://csl.sublevel3.org>
 * Distributed under the BSD License
 *
 * Compilation hint:
 * g++ test.cpp -I/sw/include -L/sw/lib -ljpeg
 *
 * $Id$
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <memory.h>
#include <math.h>
#include <jpeglib.h>

bool verbose = false;
bool color = false;
int width = 80;
int height = 25;

const char* license = "Copyright (C) 2006 Christian Stigen Larsen.\nDistributed under the BSD license";

// Printable ASCII characters, sorted least intensive to most intensive
const char* origascii = "   ...',;:clodxkO0KXNWM";

char ascii[1024];

int test_decompress(const char* file);

void help() {
	fprintf(stderr, "Usage: jp2afilename[s].jpg\n\n");
	fprintf(stderr, "Simple JPEG to ASCII converter\n\n");
	fprintf(stderr, "OPTIONS\n");
	fprintf(stderr, "    -h, --help       Print program help\n");
	fprintf(stderr, "    --size=WxH       Set output ASCII width and height, default is %dx%d\n", width, height);
	fprintf(stderr, "    --chars=...      Select character palette used to paint the image.  Leftmost character\n");
	fprintf(stderr, "                     corresponds to black pixel, rightmost to white pixel.\n");
	fprintf(stderr, "    -v, --verbose    Verbose output\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "EXAMPLE\n");
	fprintf(stderr, "   jpg2ascii --size=80x25 --chars='...oooxx@@' somefile.jpg\n\n");
	fprintf(stderr, "%s\n", license);
	exit(1);
}

void parse_options(int argc, char** argv) {
	strcpy(ascii, origascii);

	for ( int n=1; n<argc; ++n ) {
		if ( argv[n][0] != '-')
			continue;

		int hits = 0;

		if ( !strcmp(argv[n], "-h") || !strcmp(argv[n], "--help") ) help();
		if ( !strcmp(argv[n], "-v") || !strcmp(argv[n], "--verbose") ) {
			++hits;
			verbose = true;
		}

		hits += sscanf(argv[n], "--size=%dx%d", &width, &height);
		hits += sscanf(argv[n], "--chars=%1000s", ascii);

		if ( hits == 0 ) {
			fprintf(stderr, "Unknown option %s\n\n", argv[n]);
			help();
		}
	}

	// check options
	if ( strlen(ascii) == 0 )
		strcpy(ascii, origascii);

	if ( width <= 1 ) width = 1;
	if ( height <= 1 ) height = 1;
}

void print(double* accum, int width, int sizeAscii) {
	for ( int n=0; n < width; ++n ) {
		int pos = static_cast<int>(sizeAscii * accum[n]);

		// the following should never happen
		if ( pos < 0 ) pos = 0;
		if ( pos > sizeAscii ) pos = sizeAscii;

		putc(ascii[pos], stdout);
	}

	putc('\n', stdout);
}

void invert(double* accum, int width) {
	for ( int n=0; n<width; ++n )
		accum[n] = 1.0f - accum[n];
}

void clear(double* accum, int width) {
	for ( int n=0; n<width; ++n )
		accum[n] = 0.0f;
}


int main(int argc, char** argv) {
	if ( argc<2 )
		help();

	parse_options(argc, argv);

	for ( int n=1; n<argc; ++n ) {
		if ( argv[n][0] != '-')
			test_decompress(argv[n]);
	}

	return 0;
}

void normalize(double* accum, int width, double factor) {
	for ( int n=0; n<width; ++n )
		accum[n] /= factor;
}

int test_decompress(const char* file) {
	FILE * infile;

        if ( (infile = fopen(file, "rb")) == NULL ) {
            fprintf(stderr, "can't open %s\n", file);
            return 1;
        }

        struct jpeg_error_mgr jerr;
	struct jpeg_decompress_struct cinfo;

        cinfo.err = jpeg_std_error(&jerr);

        jpeg_create_decompress(&cinfo);
        jpeg_stdio_src(&cinfo, infile);
	jpeg_read_header(&cinfo, TRUE);

	jpeg_start_decompress(&cinfo);

	int row_stride = cinfo.output_width * cinfo.output_components;
	JSAMPARRAY buffer = (*cinfo.mem->alloc_sarray) ((j_common_ptr) &cinfo, JPOOL_IMAGE, row_stride, 1);

	double accum[width];

	int sizeAscii = strlen(ascii) - 1;
	int comps = cinfo.out_color_components;

	int pixelsPerChar = row_stride / (comps * width);
	if ( pixelsPerChar <= 0 ) pixelsPerChar = 1;

	int linesToAdd = cinfo.output_height / height;
	if ( linesToAdd <= 0 ) linesToAdd = 1;

	if ( verbose ) {
		printf("Filename: %s\n", file);
		printf("\n");
		printf("Output width : %d\n", width);
		printf("Output height: %d\n", height);
		printf("\n");
		printf("Source width : %d\n", cinfo.output_width);
		printf("Source height: %d\n", cinfo.output_height);
		printf("\n");
		printf("ASCII characters used for printing: %d\n", 1 + sizeAscii);
		printf("Pixels per character: %d\n", pixelsPerChar);
		printf("Lines per character : %d\n", linesToAdd);
		printf("Color components    : %d\n", comps);
		printf("\n");
	}

	int linesAdded = 0;
	clear(accum, width);

	while ( cinfo.output_scanline < cinfo.output_height ) {
		jpeg_read_scanlines(&cinfo, buffer, 1);

		int currChar = 0;
		int pixelsAdded = 0;

		for ( int pixel=0; pixel < (row_stride - comps); pixel += comps) {

			for ( int addit=0; addit<comps; ++addit )
				accum[currChar] += static_cast<double>( buffer[0][pixel + addit] / (comps * 255.0f) );

			++pixelsAdded;

			if ( pixelsAdded >= pixelsPerChar ) {
				if ( currChar>=width )
					break;

				pixelsAdded = 0;
				++currChar;
			}
		}

		++linesAdded;

		if ( linesAdded > linesToAdd ) {
			normalize(accum, width, pixelsAdded * linesAdded);
			invert(accum, width);
			print(accum, width, sizeAscii);
			clear(accum, width);
			linesAdded = 0;
		} else {
			// last line? print it
			if ( (1 + cinfo.output_scanline) == cinfo.output_height ) {
				normalize(accum, width, pixelsAdded * linesAdded);
				invert(accum, width);
				print(accum, width, sizeAscii);
				clear(accum, width);
			}
		}

	}

	jpeg_finish_decompress(&cinfo);
	jpeg_destroy_decompress(&cinfo);
	fclose(infile);

	return 0;
}
